File: 494-target-sum/494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        
        @lru_cache(None)
        def dp(index, total):
            if index >= len(nums):
                return 1 if total == target else 0

            return dp(index + 1, total + nums[index]) + dp(index + 1, total - nums[index])
        
        return dp(0, 0)

File: 494-target-sum/494-target-sum.c
int total(int * nums, int numsSize){
    int res = 0;
    for(int i = 0; i < numsSize; i++)
        res += nums[i];
    return res;
}


int findTargetSumWays(int* nums, int numsSize, int target){
    int sum = total(nums, numsSize);
    int totalSums = 2 * sum + 1;
    if(target < -sum || sum < target)
        return 0;
    
    int * dp = (int *)calloc(totalSums, sizeof(int));
    dp[sum] = 1;
    
    for(int i = 0; i < numsSize; i++){
        int * allSum = (int *)calloc(totalSums, sizeof(int));
        for(int j = 0; j < totalSums; j++){
            if(dp[j] != 0){
                allSum[j + nums[i]] += dp[j];
                allSum[j - nums[i]] += dp[j];
            }
        }
        dp = allSum;
    }
    
    return dp[sum + target];
}